File: 121. Best Time to Buy and Sell Stock.cpp
class Solution {
public:
// // O(N2): TLE, didn't pass the large data case
    // int maxProfit(vector<int>& prices) {
    //     int profit = 0;
    //     for(int i = 0; i < prices.size(); i++){
    //         for(int j = i + 1; j < prices.size(); j++){
    //             if(prices[j] - prices[i] > profit)
    //                 profit = prices[j] - prices[i];
    //         }
    //     }
    //     return profit;
    // }

    // brute force: O(N)time, O(1)space, 44.31%, ok
    int maxProfit(vector<int>& prices) {
        if(prices.empty()) return 0;
        int minP = prices[0], ret = INT_MIN;
        for(int i = 1; i < prices.size(); i++){
            ret = max(ret, prices[i] - minP);
            minP = min(minP, prices[i]);
        }
        return ret > 0? ret: 0;
    }
};

File: 121. Best Time to Buy and Sell Stock.py
class Solution(object):
    
    
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        stack, ret = [], 0
        for p in prices:
            while stack and stack[-1] > p:
                stack.pop()
            if stack == []:
                stack.append(p)
            else:
                ret = max(ret, p - stack[-1])
        return ret
    
    # record the minimal val & the maximum profit sofar: O(N) time, O(1) space
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        minPrice, maxProfit = float('inf'), 0
        for i in xrange(len(prices)):
            minPrice = min(minPrice, prices[i])
            maxProfit = max(maxProfit, prices[i] - minPrice)
        return maxProfit
                
            
                
                
        